main.tex

```diff
@@ -7,28 +7,45 @@ \section{Complex Predicates~vs.~Complex Objects}
 
 % https://github.com/DeMaCS-UNICAL/Angry-HEX/blob/4ea72273518dfde6240ece9dadea927fd1b79c3c/dlv/tactic.dlv#L133
 
-$$ canPush(o_a,o_b) :- \&canPush[objects,hills](o_a,o_b). $$
+$$ canPush(o_a,o_b) :- \&canPush[objects,hills](o_a,o_b). \label{main:rule-1} $$
 
-Intuitively, $canPush(o_a, o_b)$ should be true if the object $o_a$ can push the object $o_b$ (which requires some minimum distance and geometric range).
+Intuitively, $canPush(o_a, o_b)$ should be true if the object $o_a$ can push the object $o_b$.
+To establish the truth of $canPush(o_a, o_b)$ spatial information, like minimum distance and geometric range of the objects, is required; this is obtained by outsourcing the computation to an external source, by means of the external predicate $\&canPush$.
+The required knowledge is given as input to $\&canPush$ by using the predicates $objects$ and $hills$: indeed, $\&canPush$ is a \emph{higher-order predicate}.
 
-Note that $objects$ and $hills$ are predicates, and consequently $\&canPush$ is a higher order predicate. We observe that $objects$ and $hills$ characterize the set of objects present in the current level of the game and the structure of the ground in the virtual environment respectively.
+A useful observation is that the predicates $objects$ and $hills$ both represent lists: $objects$ characterizes the set of objects that are present in the current level, while $hills$ encodes the the morphological structure of the ground of the current level.
```

Comment:
On what ground do you argue that they represent lists? AFAIK no order is guaranteed.